File: src/security/patterns/python.ts
import { SecurityPattern, VulnerabilityType } from '../types.js';

/**
 * Python-specific security patterns for vulnerability detection
 */
export const pythonSecurityPatterns: SecurityPattern[] = [
  // SQL Injection patterns
  {
    id: 'python-sql-injection-format',
    type: VulnerabilityType.SQL_INJECTION,
    pattern: /execute\s*\(\s*["`'].*%s.*["`']\s*%\s*\w+/g,
    severity: 'high',
    message: 'SQL injection vulnerability: Using string formatting in SQL queries',
    remediation: 'Use parameterized queries with execute() method parameters',
    cwe: 'CWE-89',
    owasp: 'A03',
    language: 'python'
  },
  {
    id: 'python-sql-injection-fstring',
    type: VulnerabilityType.SQL_INJECTION,
    pattern: /execute\s*\(\s*f["'`].*\{.*\}.*["'`]/g,
    severity: 'high',
    message: 'SQL injection vulnerability: Using f-strings in SQL queries',
    remediation: 'Use parameterized queries instead of f-string formatting',
    cwe: 'CWE-89',
    owasp: 'A03',
    language: 'python'
  },
  {
    id: 'python-sql-injection-concat',
    type: VulnerabilityType.SQL_INJECTION,
    pattern: /execute\s*\(\s*["'`].*["'`]\s*\+\s*\w+/g,
    severity: 'high',
    message: 'SQL injection vulnerability: String concatenation in SQL queries',
    remediation: 'Use parameterized queries with execute() method parameters',
    cwe: 'CWE-89',
    owasp: 'A03',
    language: 'python'
  },

  // Command Injection patterns
  {
    id: 'python-command-injection-os-system',
    type: VulnerabilityType.COMMAND_INJECTION,
    pattern: /os\.system\s*\(\s*.*\+|os\.system\s*\(\s*f["'`].*\{/g,
    severity: 'critical',
    message: 'Command injection vulnerability: Unsanitized input in os.system()',
    remediation: 'Use subprocess.run() with shell=False and validate inputs',
    cwe: 'CWE-78',
    owasp: 'A03',
    language: 'python'
  },
  {
    id: 'python-command-injection-subprocess-shell',
    type: VulnerabilityType.COMMAND_INJECTION,
    pattern: /subprocess\.(run|call|check_call|Popen)\s*\([^)]*shell\s*=\s*True/g,
    severity: 'high',
    message: 'Command injection vulnerability: subprocess with shell=True',
    remediation: 'Use shell=False and pass command as list of arguments',
    cwe: 'CWE-78',
    owasp: 'A03',
    language: 'python'
  },

  // Deserialization vulnerabilities
  {
    id: 'python-unsafe-pickle',
    type: VulnerabilityType.INSECURE_DESERIALIZATION,
    pattern: /pickle\.(loads?|load)\s*\(/g,
    severity: 'critical',
    message: 'Insecure deserialization: pickle.loads() can execute arbitrary code',
    remediation: 'Use json.loads() or implement custom deserialization with validation',
    cwe: 'CWE-502',
    owasp: 'A08',
    language: 'python'
  },
  {
    id: 'python-unsafe-eval',
    type: VulnerabilityType.INSECURE_DESERIALIZATION,
    pattern: /\beval\s*\(/g,
    severity: 'critical',
    message: 'Code injection vulnerability: eval() can execute arbitrary code',
    remediation: 'Use ast.literal_eval() for safe evaluation of literals',
    cwe: 'CWE-95',
    owasp: 'A03',
    language: 'python'
  },

  // Path Traversal
  {
    id: 'python-path-traversal-open',
    type: VulnerabilityType.PATH_TRAVERSAL,
    pattern: /open\s*\(\s*.*\+|open\s*\(\s*f["'`].*\{/g,
    severity: 'medium',
    message: 'Path traversal vulnerability: Unsanitized file paths in open()',
    remediation: 'Validate and sanitize file paths, use os.path.join() safely',
    cwe: 'CWE-22',
    owasp: 'A01',
    language: 'python'
  },

  // Weak cryptography
  {
    id: 'python-weak-hash-md5',
    type: VulnerabilityType.WEAK_CRYPTOGRAPHY,
    pattern: /hashlib\.md5\s*\(/g,
    severity: 'medium',
    message: 'Weak cryptography: MD5 is cryptographically broken',
    remediation: 'Use SHA-256 or SHA-3 for cryptographic hashing',
    cwe: 'CWE-327',
    owasp: 'A02',
    language: 'python'
  },
  {
    id: 'python-weak-hash-sha1',
    type: VulnerabilityType.WEAK_CRYPTOGRAPHY,
    pattern: /hashlib\.sha1\s*\(/g,
    severity: 'medium',
    message: 'Weak cryptography: SHA-1 is deprecated for security purposes',
    remediation: 'Use SHA-256 or SHA-3 for cryptographic hashing',
    cwe: 'CWE-327',
    owasp: 'A02',
    language: 'python'
  },

  // Debug/Development issues
  {
    id: 'python-debug-true',
    type: VulnerabilityType.DEBUG_MODE,
    pattern: /DEBUG\s*=\s*True/g,
    severity: 'medium',
    message: 'Security misconfiguration: Debug mode enabled in production',
    remediation: 'Set DEBUG = False in production environments',
    cwe: 'CWE-489',
    owasp: 'A05',
    language: 'python'
  },

  // YAML deserialization
  {
    id: 'python-unsafe-yaml-load',
    type: VulnerabilityType.INSECURE_DESERIALIZATION,
    pattern: /yaml\.load\s*\([^)]*(?!Loader\s*=\s*yaml\.SafeLoader)/g,
    severity: 'critical',
    message: 'Insecure deserialization: yaml.load() without SafeLoader',
    remediation: 'Use yaml.safe_load() or yaml.load() with SafeLoader',
    cwe: 'CWE-502',
    owasp: 'A08',
    language: 'python'
  }
];

File: lib/rsolv/security/patterns/python.ex
defmodule Rsolv.Security.Patterns.Python do
  @moduledoc """
  Python security patterns for detecting vulnerabilities.

  This module contains 12 security patterns specifically designed for Python
  code. Each pattern includes detection rules, test cases, and educational
  documentation.
  """

  # Import new pattern modules
  alias Rsolv.Security.Patterns.Python.{
    UnsafePickle,
    UnsafeEval,
    SqlInjectionFormat,
    SqlInjectionFstring,
    SqlInjectionConcat,
    CommandInjectionOsSystem,
    CommandInjectionSubprocessShell,
    PathTraversalOpen,
    WeakHashMd5,
    WeakHashSha1,
    DebugTrue,
    UnsafeYamlLoad
  }

  @doc """
  Returns all Python security patterns.

  ## Examples

      iex> patterns = Rsolv.Security.Patterns.Python.all()
      iex> length(patterns)
      12
      iex> Enum.all?(patterns, &match?(%Rsolv.Security.Pattern{}, &1))
      true
  """
  def all do
    [
      sql_injection_format(),
      sql_injection_fstring(),
      sql_injection_concat(),
      command_injection_os_system(),
      command_injection_subprocess_shell(),
      unsafe_pickle(),
      unsafe_eval(),
      path_traversal_open(),
      weak_hash_md5(),
      weak_hash_sha1(),
      debug_true(),
      unsafe_yaml_load()
    ]
  end

  @doc """
  SQL Injection via string formatting pattern.

  Detects SQL queries using % string formatting with user input.

  ## Examples

      iex> pattern = Rsolv.Security.Patterns.Python.sql_injection_format()
      iex> pattern.id
      "python-sql-injection-format"
      iex> pattern.severity
      :high
      
  ## Detection Examples

      iex> pattern = Rsolv.Security.Patterns.Python.sql_injection_format()
      iex> vulnerable = ~S|cursor.execute("SELECT * FROM users WHERE id = %s" % user_id)|
      iex> Regex.match?(pattern.regex, vulnerable)
      true
      iex> safe = ~S|cursor.execute("SELECT * FROM users WHERE id = %s", (user_id,))|
      iex> Regex.match?(pattern.regex, safe)
      false
  """
  def sql_injection_format do
    SqlInjectionFormat.pattern()
  end

  @doc """
  SQL Injection via f-string formatting pattern.

  Detects SQL queries using f-strings with user input.

  ## Examples

      iex> pattern = Rsolv.Security.Patterns.Python.sql_injection_fstring()
      iex> pattern.id
      "python-sql-injection-fstring"
      iex> pattern.severity
      :high
      
  ## Detection Examples

      iex> pattern = Rsolv.Security.Patterns.Python.sql_injection_fstring()
      iex> vulnerable = ~S|cursor.execute(f"SELECT * FROM users WHERE name = '{name}'")|
      iex> Regex.match?(pattern.regex, vulnerable)
      true
      iex> safe = ~S|cursor.execute("SELECT * FROM users WHERE name = %s", (name,))|
      iex> Regex.match?(pattern.regex, safe)
      false
  """
  def sql_injection_fstring do
    SqlInjectionFstring.pattern()
  end

  @doc """
  SQL Injection via string concatenation pattern.

  Detects SQL queries built using string concatenation.

  ## Examples

      iex> pattern = Rsolv.Security.Patterns.Python.sql_injection_concat()
      iex> vulnerable = ~S|cursor.execute("SELECT * FROM users WHERE id = " + user_id)|
      iex> Regex.match?(pattern.regex, vulnerable)
      true
  """
  def sql_injection_concat do
    SqlInjectionConcat.pattern()
  end

  @doc """
  Command Injection via os.system pattern.

  Detects command execution using os.system with user input.

  ## Examples

      iex> pattern = Rsolv.Security.Patterns.Python.command_injection_os_system()
      iex> vulnerable = ~S|os.system("ls " + user_input)|
      iex> Regex.match?(pattern.regex, vulnerable)
      true
  """
  def command_injection_os_system do
    CommandInjectionOsSystem.pattern()
  end

  @doc """
  Command Injection via subprocess with shell=True pattern.

  Detects subprocess calls with shell=True.

  ## Examples

      iex> pattern = Rsolv.Security.Patterns.Python.command_injection_subprocess_shell()
      iex> vulnerable = ~S|subprocess.run(cmd, shell=True)|
      iex> Regex.match?(pattern.regex, vulnerable)
      true
  """
  def command_injection_subprocess_shell do
    CommandInjectionSubprocessShell.pattern()
  end

  @doc """
  Insecure Deserialization via pickle pattern.

  Detects usage of pickle.loads/load which can execute arbitrary code.

  ## Examples

      iex> pattern = Rsolv.Security.Patterns.Python.unsafe_pickle()
      iex> vulnerable = ~S|data = pickle.loads(user_data)|
      iex> Regex.match?(pattern.regex, vulnerable)
      true
  """
  def unsafe_pickle do
    UnsafePickle.pattern()
  end

  @doc """
  Code Injection via eval() pattern.

  Detects usage of eval() which can execute arbitrary code.

  ## Examples

      iex> pattern = Rsolv.Security.Patterns.Python.unsafe_eval()
      iex> vulnerable = ~S|result = eval(user_input)|
      iex> Regex.match?(pattern.regex, vulnerable)
      true
  """
  def unsafe_eval do
    UnsafeEval.pattern()
  end

  @doc """
  Path Traversal via open() pattern.

  Detects file operations with unsanitized paths.

  ## Examples

      iex> pattern = Rsolv.Security.Patterns.Python.path_traversal_open()
      iex> vulnerable = ~S|open("/uploads/" + filename)|
      iex> Regex.match?(pattern.regex, vulnerable)
      true
  """
  def path_traversal_open do
    PathTraversalOpen.pattern()
  end

  @doc """
  Weak Cryptography using MD5 pattern.

  Detects usage of MD5 for cryptographic purposes.

  ## Examples

      iex> pattern = Rsolv.Security.Patterns.Python.weak_hash_md5()
      iex> vulnerable = ~S|hashlib.md5(password.encode())|
      iex> Regex.match?(pattern.regex, vulnerable)
      true
  """
  def weak_hash_md5 do
    WeakHashMd5.pattern()
  end

  @doc """
  Weak Cryptography using SHA1 pattern.

  Detects usage of SHA1 for security purposes.

  ## Examples

      iex> pattern = Rsolv.Security.Patterns.Python.weak_hash_sha1()
      iex> vulnerable = ~S|hashlib.sha1(data)|
      iex> Regex.match?(pattern.regex, vulnerable)
      true
  """
  def weak_hash_sha1 do
    WeakHashSha1.pattern()
  end

  @doc """
  Debug Mode Enabled pattern.

  Detects DEBUG = True in Python code (common in Django).

  ## Examples

      iex> pattern = Rsolv.Security.Patterns.Python.debug_true()
      iex> vulnerable = ~S|DEBUG = True|
      iex> Regex.match?(pattern.regex, vulnerable)
      true
  """
  def debug_true do
    DebugTrue.pattern()
  end

  @doc """
  Unsafe YAML Loading pattern.

  Detects yaml.load() without SafeLoader.

  ## Examples

      iex> pattern = Rsolv.Security.Patterns.Python.unsafe_yaml_load()
      iex> vulnerable = ~S|data = yaml.load(user_input)|
      iex> Regex.match?(pattern.regex, vulnerable)
      true
  """
  def unsafe_yaml_load do
    UnsafeYamlLoad.pattern()
  end
end
